pkg/referrer/referrer.go

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package referrer
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/containerd/nydus-snapshotter/pkg/auth"
+	"github.com/containerd/nydus-snapshotter/pkg/label"
+
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+const metadataNameInLayer = "image/image.boot"
+
+type referrer struct {
+	// The resolver is used for image pull or fetches requests. The best practice
+	// in containerd is that each resolver instance is used only once for a request
+	// and is destroyed when the request completes. When a registry token expires,
+	// the resolver does not re-apply for a new token, so it's better to create a
+	// new resolver instance using resolverFunc for each request.
+	resolverFunc  func(plainHTTP bool) remotes.Resolver
+	WithPlainHTTP bool
+}
+
+func newReferrer(keyChain *auth.PassKeyChain, insecure bool) *referrer {
+	// nolint:unparam
+	credFunc := func(string) (string, string, error) {
+		if keyChain == nil {
+			return "", "", nil
+		}
+		return keyChain.Username, keyChain.Password, nil
+	}
+
+	newClient := func(insecure bool) *http.Client {
+		client := http.DefaultClient
+		transport := http.DefaultTransport.(*http.Transport)
+		transport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: insecure,
```

Comment:
Should default to secured transport.

Reply:
Updated, should follow with the `SkipVerify` option.